### PR TITLE
Add default community health files

### PR DESCRIPTION
**Pull request by HiddenBug**
This PR adds default community health files for the organization (like described [here](https://help.github.com/en/github/building-a-strong-community/creating-a-default-community-health-file-for-your-organization)) which will be used for any public repository of the organization.

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+```
+dpkg -l ros-melodic-pilz*
+dpkg -l ros-melodic-prbt*
+```
+  
+**Note:**  
+If you are using `kinetic`, please substitute `melodic` with `kinetic`.
+  
+If you are building from source, tell us for each PILZ-ROS
+repository you are building, which version/commit id you are using.
```
Comment (agutenkunst) [inline CONTRIBUTING.md]:
```suggestion
repository you are building, which version/commit you are using.  
```

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+  
+If you are building from source, tell us for each PILZ-ROS
+repository you are building, which version/commit id you are using.  
+
+### Steps to reproduce bug
+The step by step description, enables us to easily reproduce the bug with 
+our hardware which, in turn, is key in fixing the bug.
+Try to be as specific as possible. If it possible, try to reduce the scenario 
+which produces the bug, to the absolute minimum. 
+In other words, try to minimize the number of bugs,
```
Comment (agutenkunst) [inline CONTRIBUTING.md]:
```suggestion
In other words, try to minimize the number of steps,
```

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+needed to reproduce the bug.
+
+### Expected behavior
+Tell us what you would expect to happen. This information allows us to verify,
+if it is an actual bug or maybe just a misunderstand of some functionality.
```
Comment (agutenkunst) [inline CONTRIBUTING.md]:
```suggestion
if it is an actual bug or maybe just a misunderstanding of some functionality.
```

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+logging output, to ensure that the logging output format remains intact.
+
+### Reference to test
+If possible you should provide us with a test, which allows us 
+to reproduce the bug and check if the bug is fixed.
+
+## Adding features/ fixing things
+If you want to develop new features or fix a bug, you should first have a look
+at our [development guidelines](development_guidelines.md).
+
```
Comment (agutenkunst) [inline CONTRIBUTING.md]:
```suggestion

If you have only a rough idea that you want to discuss, or if you are not sure about the best way to implement the feature we encourage you to open a [draft pull request](https://github.blog/2019-02-14-introducing-draft-pull-requests/) as a basis for discussion.
```

File: development_guidelines.md [hunk 5]
```diff
@@ -0,0 +1,415 @@
...
+do so by forking one of our repositories, make changes, and then create a pull
+request on our repository.  
+To check that the continuous integration runs without problems on 
+*your* forked repository, you have to activate Travis for your repository.
+For more information on how to do that have a look at the
+[Travis tutorials](https://docs.travis-ci.com/user/tutorial/#to-get-started-with-travis-ci).  
+  
+**Note:**  
+In our repositories already exists a `.travis.yml` file, so there is no need
+to create your own.
```
Comment (agutenkunst) [inline development_guidelines.md]: Remove?

File: development_guidelines.md [hunk 6]
```diff
@@ -0,0 +1,415 @@
...
+you see that our CI is based on the 
+[ROS industrial CI configurations](https://github.com/ros-industrial/industrial_ci)
+for ROS. In some cases, for example in case of errors only appearing on 
+Travis, etc., it might be advantageous to do *exactly* what Travis does locally 
+(At least in terms of building the packages and executing the tests).
+To achieve this, have a look at the explanations given in the 
+[ros-industrial repository](https://github.com/ros-industrial/industrial_ci/blob/legacy/doc/index.rst#run-industrial_ci-on-local-host).
+
+## Code style guide
+<!---TODO: Are we really compliant with the ROS coding style guide? --->
```
Comment (agutenkunst) [inline development_guidelines.md]: Are we?

File: development_guidelines.md [hunk 7]
```diff
@@ -0,0 +1,415 @@
...
+  - The architecture diagrams are up to date and easily understandable.
+- Code:
+  - The code has to comply with the clean code rules 
+(see "Clean Code: A Handbook of Agile Software Craftsmanship" by Robert C. Martin).
+- Tests:
+  - The tests are easy to understand (e.g. each test only checks one aspect 
+of a component, etc.)
+  - The tests are comprehensively documented (how-to document tests see
+[here](#how-to-document-tests))
+
```
Comment (agutenkunst) [inline development_guidelines.md]: - Documentation
  - The creation of the documentation via rosdoc_lite shows no errors

File: development_guidelines.md [hunk 8]
```diff
@@ -0,0 +1,415 @@
...
+Python code, a tool named [rosdoc_lite](http://wiki.ros.org/rosdoc_lite)
+is used:
+- How to install the tool see the ROS wiki:
+[rosdoc_lite Installation](http://wiki.ros.org/rosdoc_lite#Installation). 
+- How to use rosdoc_lite see the ROS wiki:
+ [Generate document locally](http://wiki.ros.org/rosdoc_lite#Generate_document_locally).
+
+## Testing
+
+### How-to document tests
```
Comment (agutenkunst) [inline development_guidelines.md]: Now that we gained some experience on this I would really like to re-discuss this step by step test documentation. Up for it?

**Comment by jschleicher** [comment]
General note: organization-wide rules may also affect @tgodau et al.

**Comment by jschleicher** [comment]
Regarding the code-of-coduct please refer to the recent discussion on https://github.com/ros-planning/moveit/pull/1684, which made some valid points and have choosen a concise text from the django project now (I like that).

**Comment by HiddenBug** [comment]
> Regarding the code-of-coduct please refer to the recent discussion on [ros-planning/moveit#1684](https://github.com/ros-planning/moveit/pull/1684), which made some valid points and have choosen a concise text from the django project now (I like that).

OK. I sense that there might be the need to discuss this in more detail. But please note, that I cannot choose a certain code-of-conduct because one developer likes it. If I do this, I will change the code-of-conduct on a daily basis. I suggest to remove the code-of-conduct from this PR and open another PR which deals with the code-of-conduct discussion. Is this ok for you?

**Comment by HiddenBug** [comment]
> General note: organization-wide rules may also affect @tgodau et al.
> So, maybe a review by BPC software would be appropriate?

Yes, I think we can trigger a review by the BPC software. But as far as I can see @tgodau does not have any public repositories. Therefore, it should have no relevance for him right now. If it does, he is free to comment on this PR.

**Comment by HiddenBug** [comment]
> General note: organization-wide rules may also affect @tgodau et al.
> So, maybe a review by BPC software would be appropriate?

@jschleicher Please also note, that each public-repository can still define it's own rules. These are just the default rules which are used if the public-repository itself does not define any.

File: CONTRIBUTING.md (hunk 1 above)
Comment (HiddenBug) [reply CONTRIBUTING.md]: Done.

File: CONTRIBUTING.md (hunk 2 above)
Comment (HiddenBug) [reply CONTRIBUTING.md]: Done.

File: CONTRIBUTING.md (hunk 4 above)
Comment (HiddenBug) [reply CONTRIBUTING.md]: Done.

File: CONTRIBUTING.md (hunk 3 above)
Comment (HiddenBug) [reply CONTRIBUTING.md]: Done.

File: development_guidelines.md (hunk 5 above)
Comment (HiddenBug) [reply development_guidelines.md]: Done.

File: development_guidelines.md (hunk 6 above)
Comment (HiddenBug) [reply development_guidelines.md]: Don't know. But at one time we said we want too. I simply deleted the TODO :) -> Done.

File: development_guidelines.md (hunk 7 above)
Comment (HiddenBug) [reply development_guidelines.md]: Currently, rosdoc_lite has errors not related to us. Until these errors are fixed, I'm against this statement.

Comment (agutenkunst) [reply development_guidelines.md]: - Documentation
  - The creation of the documentation via rosdoc_lite shows no errors related to your code

?

File: development_guidelines.md (hunk 8 above)
Comment (agutenkunst) [reply development_guidelines.md]: I get your points by I figure that all of these points apply not only to test code but for production code as well.
While one can argue against some difference in rules for tests (e.g. DAMP vs DRY (see [link](https://stackoverflow.com/a/11837973)).

However when it comes to documentation imho Clean Code - Chapter 4,
, https://twitter.com/unclebobmartin/status/870311898545258497?lang=en seems pretty compelling.
There are reasons for comments in unittests (see https://blog.cleancoder.com/uncle-bob/2017/02/23/NecessaryComments.html) however they should originate from a lack of expressing the "why" not the "what". ( [https://blog.codinghorror.com/code-tells-you-how-comments-tell-you-why/](https://blog.codinghorror.com/code-tells-you-how-comments-tell-you-why/) )

> In this case, the test code has to explain itself. However, this approach does only work if I'm disciplined enough to always ensure that my test are easy to read and understand. But what happens for example if somebody refactors the test and does not ensure that the readability is maintained (This also happened before.)?

The documentation scheme does not help against a lack of discipline here. Let's say for the sake of the argument the code is written in a fast rather "as long as it works" fashion. Is it reasonable to assume that the same person would suddenly start to write a total expressiv, correct and well-thought comment above the questionable code?
Same holds for refactorings as well if not even more.

>Also: How do we define the term "readability" or "understandability" when it comes to test code?

I see no reason why this should be different from production code. Surely these are rather subjective terms and since these code qualities are completely irrelevant for the machine (where compilability is enough :wink:) since they address other programmers (including yourself in a few weeks :neckbeard:)

>Furthermore, this approach misses a short summary (one sentence) explaining what the test intents to do. This information is something that even very "good code" cannot express that easily.

If it is possible the name should indicate summarise "what".
(I also find [https://osherove.com/blog/2005/4/3/naming-standards-for-unit-tests.html](https://osherove.com/blog/2005/4/3/naming-standards-for-unit-tests.html) interesting)

>Even if the programmer knows the documentation strategy: Who guarantees that the programmer does update the documentation?

Maybe my mistake was that I only expressed what I disliked. So what I don't want is another "documentation strategy" but rather a precise usage of comments only at points where the reasoning cannot be expressed in any other way. (see again https://blog.cleancoder.com/uncle-bob/2017/02/23/NecessaryComments.html)

>Provide documentation scheme:
This is basically the current approach. Like @agutenkunst mentioned there are some issues with this scheme. Looking at the alternative approaches, however, we see that the issues of this approach are not so different from the other approaches: What happens if somebody refactors a test? Who guarantees that the documentation is updated?

Could you provide some references to alternativ approaches or to our current approach?

>I don't think that the current documentation scheme itself is the problem but rather our inability to automatically check the actuality of the documentation and the readability of the code.

This is something which will ~never~ only be possible by AI in a few years :stuck_out_tongue_winking_eye:
Tbh and excuse me if I am wrong: But did we maybe by introducing the current scheme try to create something measurable? (We could also check the scheme in the CI - just kidding - please don't)

>We can of course reduce the documentation overhead in order to reduce the amount of work one has to do if a test is refactored.

Totally up for it.

>But again: How do we define readability and understandability when it comes to testing? And: Who checks that the readability and understandability are maintained or exist in the first place?

I know this sounds naiv but: This should be checked by the reviewer.
Also this writing code with high expressiv and readable code is not something that comes for free by just writing enough lines for a few years. I think this needs to be learned explicitly and in exchange with other programmers reading the code one wrote. Thus reviews also focussing on this seem ideal for this.

To summarise what I would desire:
- Don't make the current scheme mandatory or even remove it from the guidelines
- Focus on the readability/expressiveness of the (test-)code itself
  - Encourage programmers to engage in group discussions if they are not sure if their code has these qualities
- Focus on creating useful testhelpers which allow to abstract code lines not relevant in detail and provide domain specific Macros (e.g. EXPECT_POSITION)
 
@ct2034 @leunMar @martiniil your opinion?

**Comment by agutenkunst** [comment]
> > General note: organization-wide rules may also affect @tgodau et al.
> > So, maybe a review by BPC software would be appropriate?
> 
> Yes, I think we can trigger a review by the BPC software. But as far as I can see @tgodau does not have any public repositories. Therefore, it should have no relevance for him right now. If it does, he is free to comment on this PR.

I also figure that since this documentation and its location is nothing final since more developers will join on github and bring their own ideas and desires. Thus such as life itself these documents iare in constant flow :innocent: :innocent: :yum: - they are agile just as we are :wink: .

File: development_guidelines.md (hunk 8 above)
Comment (jschleicher) [reply development_guidelines.md]: I totally agree with keeping comments as close to / within the test (keeping-life-simple rather than "discipline" ;-)) and getting rid of the "headline" steps in the future.

File: development_guidelines.md [hunk 9]
```diff
@@ -0,0 +1,424 @@
...
+[ros-industrial repository](https://github.com/ros-industrial/industrial_ci/blob/legacy/doc/index.rst#run-industrial_ci-on-local-host).
+
+## Code style guide
+We follow the [ROS coding style guide](http://wiki.ros.org/CppStyleGuide). 
+Currently, the styling of the code is not automatically checked by the CI, 
+so please make sure that you comply with the ROS coding style guide. 
+
+### How-to format code automatically
+If you want to automatically format your code using the ROS style guide,
+then see [here](http://wiki.ros.org/CppStyleGuide#Autoformatting_of_ROS_Code).
```
Comment (jschleicher) [inline development_guidelines.md]: Please avoid (here-links)[http://stephanieleary.com/2015/05/why-click-here-is-a-terrible-link-and-what-to-write-instead/] for easier readability. Thanks.

**Comment by jschleicher** [comment]
> remove the code-of-conduct from this PR
Yep, discussing the Code-of-conduct in a separate PR sounds good.

File: development_guidelines.md (hunk 9 above)
Comment (HiddenBug) [reply development_guidelines.md]: Done.

File: development_guidelines.md (hunk 8 above)
Comment (HiddenBug) [reply development_guidelines.md]: @agutenkunst  Although, I think this is a good and useful discussion, I suggested closing this PR without modifications and then opening a new PR whose sole purpose is to continue this discussion.

Comment (agutenkunst) [reply development_guidelines.md]: How about removing the controversial part and opening a PR for integrating it?

**Comment by leunMar** [comment]
> How about removing the controversial part and opening a PR for integrating it?

> remove the code-of-conduct from this PR

:+1:

File: development_guidelines.md (hunk 8 above)
Comment (HiddenBug) [reply development_guidelines.md]: As you wish, sir. -> Done.

File: development_guidelines.md (hunk 7 above)
Comment (HiddenBug) [reply development_guidelines.md]: Remains left out -> Open new PR later on. -> Done.
